### DISCUSSION TITLE
JS Target Support with Apache Royale

### DISCUSSION THREAD

**Discussion by Fancy2209** [airsdk/Adobe-Runtime-Support#2844] [Ideas]
While AIR provides a good way for packaging AS3 Apps for Mobile and Desktop, it currently doesn't support anything related to web.

My idea is for HARMAN to provide a swc that could be used with [Apache Royale](https://royale.apache.org) , that allows the use of the AIR/Flash APIs with it, since while Royale supports transpiling AS to JS, it does not support the flash apis.

This could also be a easier way of getting AIR apps running on web than attempting to get the AIR runtime running on Web Assembly.

**Comment by Ender22** [comment]
That's interesting, my company for one would be thrilled as we have a b2b app and often businesses want a web integration that we just can't provide.

**Reply by Fancy2209** [reply]
If this doesn't end up happening, you could look into OpenFL's Royale support, I recommend using a manually built version ofthe package from the OpenFL-JS repo due to the NPM package being currently outdated and looking at the openfl-examples-as3 repository for examples on how to do use it.

**Comment by ajwfrost** [comment]
... although this has a challenge in that the SWC files used for the AIR APIs (airglobal.swc) has a lot of implementations that are marked "native". Which are then written in C++...  this is the "avmglue" that people talk about, which is what needs to be created in order to get Flash/AIR apps running on new platforms.

Royale works well because it is based on Flex and provides new 'components' from which the UI is built up. So your MXML/AS code is transpiled to JavaScript, but your UI components are written out as HTML/CSS using a normal browser DOM.

Having said that .. I guess it would be feasible to do this for some AIR APIs if they're written in AS3 with some JavaScript code (like Royale). A key issue would be rendering.. you'd have to implement all of this somehow. I'm not quite sure how it might then interact with Royale where they have all the bootstrap code that expects an MXML-based application. But the same theory could apply perhaps, where you compile AS3 into JavaScript and then add some framework to provide the application loop, event handling etc. All you'd then need is the rendering (maybe steal code from OpenFL for that?) plus whatever other API implementations are required. DisplayObject etc would need to be handled, but should be possible (but good luck trying to get layouts and positioning working with JS/CSS!)

I would add that this isn't something Harman should do, this would be better done as an open source project. The AS3 APIs are published under creative commons licenses so as far as I'm aware, it would be possible to implement new versions of these APIs without any problems (no Oracle-vs-Google type issues).

thanks

**Reply by Fancy2209** [reply]
Yeah I see, thanks for replaying
OpenFL and its Starling fork do actually have support royale, I was mostly suggesting it due to being interested on how a official take on a royale integration would be, but from what you say I see that it is more trouble than it's worth.

**Reply by Fancy2209** [reply]
Also royale doesn't require mxml from my understanding, you can use the "js" target, like how OpenFL does
